server/embed/config.go

```diff
@@ -151,7 +151,7 @@ func init() {
 type Config struct {
 	Name   string `json:"name"`
 	Dir    string `json:"data-dir"`
-	WalDir string `json:"wal-dir"`
+	WALDir string `json:"wal-dir"`
```

Comment:
It will impact applications in which etcd is embedded.

PLease let's follow the same deprecation process to rename it. This comment applies to following changes as well.

Reply:
@ahrtr, I don't think (or at least I can't think how) it's possible to deprecate a struct field; as Go doesn't support overriding the `=` operator, it would be impossible to know and reflect its value.

I can think of a couple of solutions:
1. Ignore these (`//revive:ignore ...`), and add a `TODO` to address them in the next major release
2. Deprecate, and everywhere the deprecated variable is referenced, call the non-deprecated version, i.e.
    ```diff
    -minVersion, err := tlsutil.GetTLSVersion(cfg.TLSMinVersion)
    +var tlsMinVersion string
    +if cfg.TlsMinVersion != "" {
    +  tlsMinVersion = cfg.TlsMinVersion
    +} else {
    +  tlsMinVersion = cfg.TLSMinVersion
    +}
    +minVersion, err := tlsutil.GetTLSVersion(tlsMinVersion)
    ```

However, as I mentioned, the fields won't be in sync. So, it adds the overhead of constantly checking both when referencing them. And could introduce bugs in external applications accessing these fields.

The problem with the first option is that it will likely not be addressed in the next major release.

Still, I prefer the first option. Thoughts?

Reply:
We can add a method like below,
```
// getWALDir returns the `WALDir` if not empty, otherwise `WalDir`.
// TODO: remove this method once we remove the deprecated `WalDir`.
func (c *Config) getWALDir() string {
	if len(c.WALDir) > 0 {
		return c.WALDir
	}

	return c.WalDir
}
```

The whole diff for your reference,
```
$ git diff
diff --git a/server/embed/config.go b/server/embed/config.go
index 3bbdf0c9e..6696b7f33 100644
--- a/server/embed/config.go
+++ b/server/embed/config.go
@@ -147,11 +147,24 @@ func init() {
        defaultHostname, defaultHostStatus = netutil.GetDefaultHost()
 }
 
+// getWALDir returns the `WALDir` if not empty, otherwise `WalDir`.
+// TODO: remove this method once we remove the deprecated `WalDir`.
+func (c *Config) getWALDir() string {
+       if len(c.WALDir) > 0 {
+               return c.WALDir
+       }
+
+       return c.WalDir
+}
+
 // Config holds the arguments for configuring an etcd server.
 type Config struct {
-       Name   string `json:"name"`
-       Dir    string `json:"data-dir"`
+       Name string `json:"name"`
+       Dir  string `json:"data-dir"`
+       // Deprecated: please use WALDir instead,
+       // to be decommissioned in 3.7
        WalDir string `json:"wal-dir"`
+       WALDir string `json:"wal-dir"`
 
        SnapshotCount uint64 `json:"snapshot-count"`
 
@@ -561,7 +574,7 @@ func NewConfig() *Config {
 func (cfg *Config) AddFlags(fs *flag.FlagSet) {
        // member
        fs.StringVar(&cfg.Dir, "data-dir", cfg.Dir, "Path to the data directory.")
-       fs.StringVar(&cfg.WalDir, "wal-dir", cfg.WalDir, "Path to the dedicated wal directory.")
+       fs.StringVar(&cfg.WALDir, "wal-dir", cfg.WALDir, "Path to the dedicated wal directory.")
        fs.Var(
                flags.NewUniqueURLsWithExceptions(DefaultListenPeerURLs, ""),
                "listen-peer-urls",
diff --git a/server/embed/etcd.go b/server/embed/etcd.go
index 42b9b78b9..2902f0ad9 100644
--- a/server/embed/etcd.go
+++ b/server/embed/etcd.go
@@ -171,7 +171,7 @@ func StartEtcd(inCfg *Config) (e *Etcd, err error) {
                ClientURLs:                               cfg.AdvertiseClientUrls,
                PeerURLs:                                 cfg.AdvertisePeerUrls,
                DataDir:                                  cfg.Dir,
-               DedicatedWALDir:                          cfg.WalDir,
+               DedicatedWALDir:                          cfg.getWALDir(),
                SnapshotCount:                            cfg.SnapshotCount,
                SnapshotCatchUpEntries:                   cfg.SnapshotCatchUpEntries,
                MaxSnapFiles:                             cfg.MaxSnapFiles,
@@ -320,7 +320,7 @@ func print(lg *zap.Logger, ec Config, sc config.ServerConfig, memberInitialized
                zap.Bool("member-initialized", memberInitialized),
                zap.String("name", sc.Name),
                zap.String("data-dir", sc.DataDir),
-               zap.String("wal-dir", ec.WalDir),
+               zap.String("wal-dir", ec.getWALDir()),
                zap.String("wal-dir-dedicated", sc.DedicatedWALDir),
                zap.String("member-dir", sc.MemberDir()),
                zap.Bool("force-new-cluster", sc.ForceNewCluster),
diff --git a/server/embed/util.go b/server/embed/util.go
index 269fbc80b..d2ca91e70 100644
--- a/server/embed/util.go
+++ b/server/embed/util.go
@@ -21,7 +21,7 @@ import (
 )
 
 func isMemberInitialized(cfg *Config) bool {
-       waldir := cfg.WalDir
+       waldir := cfg.getWALDir()
        if waldir == "" {
                waldir = filepath.Join(cfg.Dir, "member", "wal")
        }
```

Reply:
Done; the only issue is that Go doesn't allow setting the same JSON tag to two fields. So, I left the JSON tag set in the renamed (new) field. And because it's `go vet` that complains, I don't think I can add an ignore statement.

See:

```
% (cd server && 'go' 'vet' './...')
stderr: # go.etcd.io/etcd/server/v3/embed
stderr: # [go.etcd.io/etcd/server/v3/embed]
stderr: embed/config.go:158:2: struct field WALDir repeats json tag "wal-dir" also at config.go:156
stderr: embed/config.go:174:2: struct field MaxWALFiles repeats json tag "max-wals" also at config.go:172
stderr: embed/config.go:245:2: struct field TlsMinVersion repeats json tag "tls-min-version" also at config.go:242
stderr: embed/config.go:250:2: struct field TlsMaxVersion repeats json tag "tls-max-version" also at config.go:247
stderr: embed/config.go:455:2: struct field WALDir repeats json tag "wal-dir" also at config.go:156
stderr: embed/config.go:455:2: struct field MaxWALFiles repeats json tag "max-wals" also at config.go:172
stderr: embed/config.go:455:2: struct field TlsMinVersion repeats json tag "tls-min-version" also at config.go:242
stderr: embed/config.go:455:2: struct field TlsMaxVersion repeats json tag "tls-max-version" also at config.go:247

```

Reply:
Done

Reply:
Duh, right. What I was thinking is that when unmarshalling the field will only be at WALDir. Is that a use case? Or, it only marshals it?

Reply:
JSON will marshal all exported fields using the fieldName by default, and overridden by the JSON tag if present. The same when unmarshalling it. I think we just need to explicitly clarify this.

Reply:
Alright, I updated it. I hope that the message makes sense.